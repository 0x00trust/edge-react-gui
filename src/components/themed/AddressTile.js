// @flow

import Clipboard from '@react-native-community/clipboard'
import type { EdgeCurrencyConfig, EdgeCurrencyWallet, EdgeParsedUri, EdgeSpendTarget, EdgeTransaction } from 'edge-core-js'
import * as React from 'react'
import { TouchableWithoutFeedback, View } from 'react-native'
import FontAwesome from 'react-native-vector-icons/FontAwesome'
import { connect } from 'react-redux'

import { CURRENCY_PLUGIN_NAMES } from '../../constants/WalletAndCurrencyConstants'
import s from '../../locales/strings.js'
import { checkPubAddress } from '../../modules/FioAddress/util'
import type { RootState } from '../../reducers/RootReducer'
import type { FeeOption } from '../../reducers/scenes/SendConfirmationReducer'
import type { FioRequest } from '../../types/types'
import { AddressModal } from '../modals/AddressModal'
import { ButtonsModal } from '../modals/ButtonsModal'
import { Airship, showError } from '../services/AirshipInstance'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { EdgeText } from './EdgeText'
import { ScanTile } from './ScanTile.js'
import { ClickableText } from './ThemedButtons'
import { Tile } from './Tile.js'

type SpendInfo = {
  currencyCode?: string,
  metadata?: any,
  nativeAmount?: string,
  networkFeeOption?: FeeOption,
  customNetworkFee?: Object,
  publicAddress?: string,
  spendTargets?: EdgeSpendTarget[],
  lockInputs?: boolean,
  uniqueIdentifier?: string,
  otherParams?: Object,
  dismissAlert?: boolean,
  fioAddress?: string,
  fioPendingRequest?: FioRequest,
  isSendUsingFioAddress?: boolean,
  onBack?: () => void,
  onDone?: (error: Error | null, edgeTransaction?: EdgeTransaction) => void,
  beforeTransaction?: () => Promise<void>
}

type OwnProps = {
  coreWallet: EdgeCurrencyWallet,
  currencyCode: string,
  title: string,
  recipientAddress: string,
  onChangeAddress: (address: string, spendInfo?: SpendInfo) => Promise<void>
}
type StateProps = {
  fioPlugin: EdgeCurrencyConfig | null
}
type State = {
  clipboard: string,
  resolvingAddress: boolean
}
type Props = OwnProps & StateProps & ThemeProps

const isLegacyAddressUri = (parsedUri: EdgeParsedUri): boolean => {
  return !!parsedUri.legacyAddress
}

const isPaymentProtocolUri = (parsedUri: EdgeParsedUri): boolean => {
  // $FlowFixMe should be paymentProtocolUrl (lowercased)?
  return !!parsedUri.paymentProtocolURL && !parsedUri.publicAddress
}

const BITPAY = {
  domain: 'bitpay.com',
  merchantName: (memo: string) => {
    // Example BitPay memo
    // "Payment request for BitPay invoice DKffym7WxX6kzJ73yfYS7s for merchant Electronic Frontier Foundation"
    // eslint-disable-next-line no-unused-vars
    const [_, merchantName] = memo.split(' for merchant ')
    return merchantName
  }
}

class AddressTileComponent extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      clipboard: '',
      resolvingAddress: false
    }
  }

  componentDidMount(): void {
    this._setClipboard(this.props)
  }

  shouldContinueLegacy = async () => {
    const response = await Airship.show(bridge => (
      <ButtonsModal
        bridge={bridge}
        title={s.strings.legacy_address_modal_title}
        message={s.strings.legacy_address_modal_warning}
        buttons={{
          confirm: { label: s.strings.legacy_address_modal_continue },
          cancel: { label: s.strings.legacy_address_modal_cancel, type: 'secondary' }
        }}
      />
    ))
    if (response === 'confirm') {
      return true
    }

    return false
  }

  paymentProtocolUriReceived = async ({ paymentProtocolURL }: { paymentProtocolURL?: string }) => {
    const { coreWallet, onChangeAddress } = this.props
    try {
      if (!paymentProtocolURL) throw new Error('no paymentProtocolURL prop')
      const paymentProtocolInfo = await coreWallet.getPaymentProtocolInfo(paymentProtocolURL)
      const { domain, memo, nativeAmount, spendTargets } = paymentProtocolInfo

      const name = domain === BITPAY.domain ? BITPAY.merchantName(memo) : domain
      const notes = memo
      const spendInfo = {
        lockInputs: true,
        networkFeeOption: 'standard',
        metadata: {
          name,
          notes
        },
        nativeAmount,
        spendTargets,
        otherParams: { paymentProtocolInfo }
      }
      onChangeAddress(spendTargets.length && spendTargets[0].publicAddress ? spendTargets[0].publicAddress : '', spendInfo)
    } catch (e) {
      console.log(e)
      await Airship.show(bridge => (
        <ButtonsModal
          bridge={bridge}
          title={s.strings.scan_invalid_address_error_title}
          message={s.strings.scan_invalid_address_error_description}
          buttons={{
            ok: { label: s.strings.string_ok }
          }}
        />
      ))
    }
  }

  onChangeAddress = async (address: string) => {
    if (!address) return
    const { onChangeAddress, coreWallet, currencyCode, fioPlugin } = this.props

    this.setState({ resolvingAddress: true })
    let fioAddress
    if (fioPlugin) {
      try {
        const publicAddress = await checkPubAddress(fioPlugin, address.toLowerCase(), coreWallet.currencyInfo.currencyCode, currencyCode)
        fioAddress = address.toLowerCase()
        address = publicAddress
      } catch (e) {
        if (!e.code || e.code !== fioPlugin.currencyInfo.defaultSettings.errorCodes.INVALID_FIO_ADDRESS) {
          this.setState({ resolvingAddress: false })
          return showError(e)
        }
      }
    }
    try {
      const parsedUri: EdgeParsedUri & { paymentProtocolURL?: string } = await coreWallet.parseUri(address, currencyCode)

      this.setState({ resolvingAddress: false })

      if (isLegacyAddressUri(parsedUri)) {
        if (!(await this.shouldContinueLegacy())) return
      }

      if (isPaymentProtocolUri(parsedUri)) {
        return this.paymentProtocolUriReceived(parsedUri)
      }

      if (!parsedUri.publicAddress) {
        return showError(s.strings.scan_invalid_address_error_title)
      }

      // set address
      onChangeAddress(parsedUri.publicAddress, { fioAddress })
    } catch (e) {
      showError(`${s.strings.scan_invalid_address_error_title} ${s.strings.scan_invalid_address_error_description}`)
      this.setState({ resolvingAddress: false })
    }
  }

  _setClipboard = async props => {
    const coreWallet = props.coreWallet

    try {
      const uri = await Clipboard.getString()

      // Will throw in case uri is invalid
      await coreWallet.parseUri(uri)

      this.setState({
        clipboard: uri
      })
    } catch (e) {
      // Failure is acceptable
    }
  }

  onPasteFromClipboard = () => {
    const { clipboard } = this.state
    this.onChangeAddress(clipboard)
  }

  changeRecipient = async () => {
    const { coreWallet, currencyCode } = this.props
    const address = await Airship.show(bridge => (
      <AddressModal bridge={bridge} walletId={coreWallet.id} currencyCode={currencyCode} title={s.strings.scan_address_modal_title} />
    ))
    if (address) {
      this.onChangeAddress(address)
    }
  }

  onScan = async result => {
    if (result) {
      this.onChangeAddress(result)
    }
  }

  onAddressPress = (): void => {
    this.changeRecipient()
  }

  renderAddress() {
    const { theme, recipientAddress } = this.props
    const { resolvingAddress } = this.state
    const styles = getStyles(theme)
    if (resolvingAddress) {
      return <EdgeText style={styles.tilePlaceHolder}>{s.strings.resolving}</EdgeText>
    }

    return recipientAddress ? (
      <>
        <EdgeText ellipsizeMode="middle" numberOfLines={1} style={[styles.tileTextBottom, styles.rightSpace]}>
          {recipientAddress}
        </EdgeText>
        <FontAwesome name="edit" style={styles.editIcon} />
      </>
    ) : (
      <>
        <EdgeText style={[styles.tilePlaceHolder, styles.rightSpace]}>{s.strings.address_modal_default_header}</EdgeText>
        <FontAwesome name="edit" style={styles.editIcon} />
      </>
    )
  }

  render() {
    const { theme, title } = this.props
    const styles = getStyles(theme)
    const copyMessage = this.state.clipboard ? `${s.strings.string_paste}: ${this.state.clipboard}` : null
    return (
      <View>
        <Tile type="static" containerClass={styles.noBottomMargin} title={title} />
        <View style={styles.recipientBlock}>
          <View style={styles.tileRowParent}>
            <TouchableWithoutFeedback onPress={this.onAddressPress}>
              <View style={[styles.tileRow, styles.tileChildSideBorder]}>{this.renderAddress()}</View>
            </TouchableWithoutFeedback>
            <ScanTile onScan={this.onScan} />
          </View>
          {copyMessage && (
            <View style={styles.tileContainerButtons}>
              <ClickableText onPress={this.onPasteFromClipboard} style={styles.pasteButton}>
                <EdgeText ellipsizeMode="middle" numberOfLines={1} style={styles.tileTextBottom}>
                  {copyMessage}
                </EdgeText>
              </ClickableText>
            </View>
          )}
        </View>
      </View>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  tileContainerButtons: {
    backgroundColor: theme.tileBackground,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  tileRow: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: theme.tileBackground,
    paddingHorizontal: theme.rem(0.75),
    paddingVertical: theme.rem(0.5)
  },
  tileRowParent: {
    flexDirection: 'row',
    alignItems: 'center',
    borderColor: theme.deactivatedText,
    borderTopWidth: theme.rem(0.05),
    borderBottomWidth: theme.rem(0.05)
  },
  tileChildSideBorder: {
    borderColor: theme.deactivatedText,
    borderRightWidth: theme.rem(0.05)
  },
  tileTextBottom: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  tilePlaceHolder: {
    color: theme.deactivatedText,
    fontSize: theme.rem(1)
  },
  noBottomMargin: {
    marginBottom: 0
  },
  pasteButton: {
    width: '100%',
    backgroundColor: 'transparent',
    padding: theme.rem(0.75),
    margin: 0
  },
  rightSpace: {
    maxWidth: '90%',
    paddingRight: theme.rem(0.75)
  },
  recipientBlock: {
    marginBottom: theme.rem(0.125)
  },
  loader: {
    paddingVertical: theme.rem(0.075)
  },
  editIcon: {
    color: theme.iconTappable,
    fontSize: theme.rem(1)
  }
}))

export const AddressTile = connect((state: RootState): StateProps => {
  const { account } = state.core
  return {
    fioPlugin: account.currencyConfig ? account.currencyConfig[CURRENCY_PLUGIN_NAMES.FIO] : null
  }
})(withTheme(AddressTileComponent))
