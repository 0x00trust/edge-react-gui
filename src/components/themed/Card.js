// @flow

import * as React from 'react'
import { View } from 'react-native'

import { unpackEdges } from '../../util/edges'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'

type Props = {
  children: React.Node,
  warning?: boolean,
  marginRem?: number[] | number
}

class CardComponent extends React.PureComponent<Props & ThemeProps> {
  render() {
    const { children, warning, theme } = this.props
    const styles = getStyles(theme)

    return <View style={[styles.container, warning ? styles.warning : null, marginRem(this.props, theme)]}>{children}</View>
  }
}

function marginRem(props: Props, theme: Theme) {
  const marginRem = unpackEdges(props.marginRem || [0, 1])

  return {
    marginBottom: theme.rem(marginRem.bottom),
    marginLeft: theme.rem(marginRem.left),
    marginRight: theme.rem(marginRem.right),
    marginTop: theme.rem(marginRem.top)
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    padding: theme.rem(1),
    borderWidth: theme.cardBorder,
    borderColor: theme.cardBorderColor,
    borderRadius: theme.cardBorderRadius
  },
  warning: {
    borderColor: theme.warningIcon
  }
}))

export const Card = withTheme(CardComponent)
