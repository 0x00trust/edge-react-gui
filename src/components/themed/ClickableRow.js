// @flow

import * as React from 'react'
import { TouchableHighlight, TouchableOpacity, View } from 'react-native'

import { Gradient } from '../../modules/UI/components/Gradient/Gradient.ui'
import { unpackEdges } from '../../util/edges'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'

type Props = {
  onPress: () => void | (() => Promise<void>),
  highlight?: boolean,
  gradient?: boolean,
  children?: React.Node,
  underline?: boolean,

  marginRem?: number[] | number,
  paddingRem?: number[] | number
}

class ClickableRowComponent extends React.PureComponent<Props & ThemeProps> {
  renderContent() {
    const { gradient, children, underline, theme } = this.props
    const styles = getStyles(theme)
    const containerStyles = [styles.rowContainer, spacingStyles(this.props, theme), underline ? styles.underline : null]
    if (gradient) {
      return <Gradient style={containerStyles}>{children}</Gradient>
    }

    return <View style={containerStyles}>{children}</View>
  }

  render() {
    const { onPress, highlight, theme } = this.props

    if (highlight) {
      return (
        <TouchableHighlight onPress={onPress} underlayColor={theme.backgroundGradientLeft}>
          {this.renderContent()}
        </TouchableHighlight>
      )
    }

    return <TouchableOpacity onPress={onPress}>{this.renderContent()}</TouchableOpacity>
  }
}

function spacingStyles(props: Props, theme: Theme) {
  const marginRem = unpackEdges(props.marginRem || 0)
  const paddingRem = unpackEdges(props.paddingRem ?? [0, 1])

  return {
    marginBottom: theme.rem(marginRem.bottom),
    marginLeft: theme.rem(marginRem.left),
    marginRight: theme.rem(marginRem.right),
    marginTop: theme.rem(marginRem.top),
    paddingBottom: theme.rem(paddingRem.bottom),
    paddingLeft: theme.rem(paddingRem.left),
    paddingRight: theme.rem(paddingRem.right),
    paddingTop: theme.rem(paddingRem.top)
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  rowContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    height: theme.rem(4.25)
  },
  underline: {
    borderBottomWidth: theme.thinLineWidth,
    borderBottomColor: theme.lineDivider
  }
}))

export const ClickableRow = withTheme(ClickableRowComponent)
