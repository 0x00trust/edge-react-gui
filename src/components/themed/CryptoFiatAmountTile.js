// @flow
import { bns } from 'biggystring'
import * as React from 'react'

import { MAX_CRYPTO_AMOUNT_CHARACTERS } from '../../constants/WalletAndCurrencyConstants.js'
import { formatNumber } from '../../locales/intl.js'
import { getDisplayDenomination } from '../../selectors/DenominationSelectors'
import { useSelector } from '../../types/reactRedux'
import { DECIMAL_PRECISION } from '../../util/utils'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'
import { EdgeText } from './EdgeText'
import { FiatText } from './FiatText'
import { Tile } from './Tile.js'

type Props = {
  cryptoAmount: string,
  cryptoCurrencyCode: string,
  isoFiatCurrencyCode: string,
  maxCryptoChars?: number,
  title: string
}

export const CryptoFiatAmountTile = (props: Props) => {
  const { title, cryptoAmount, cryptoCurrencyCode, isoFiatCurrencyCode, maxCryptoChars } = props
  const styles = getStyles(useTheme())

  const { cryptoName, cryptoMultiplier } = useSelector(state => {
    const { name, multiplier } = getDisplayDenomination(state, cryptoCurrencyCode)
    return { cryptoName: name, cryptoMultiplier: multiplier }
  })
  // Default to 10 displayed chars for crypto amount
  const cryptoAmountDenom = bns.div(cryptoAmount, cryptoMultiplier, DECIMAL_PRECISION)
  const fmtCryptoAmount = formatNumber(cryptoAmountDenom, { toFixed: maxCryptoChars === undefined ? MAX_CRYPTO_AMOUNT_CHARACTERS : maxCryptoChars })
  const cryptoAmountText = `${fmtCryptoAmount} ${cryptoName} `
  // Fiat amount is always positive for this specific tile
  const absCryptoAmount = bns.abs(cryptoAmount)
  const fiatAmount = (
    <FiatText cryptoAmount={absCryptoAmount} cryptoCurrencyCode={cryptoCurrencyCode} isoFiatCurrencyCode={isoFiatCurrencyCode} parenthesisEnclosed />
  )

  return (
    <Tile type="static" title={title} contentPadding={false} style={styles.tileContainer}>
      <EdgeText>
        {cryptoAmountText}
        {fiatAmount}
      </EdgeText>
    </Tile>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  tileContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start'
  }
}))
