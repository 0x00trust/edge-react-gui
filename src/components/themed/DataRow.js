// @flow

import * as React from 'react'
import { View } from 'react-native'

import { unpackEdges } from '../../util/edges'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'

type Props = {
  image?: React.Node,
  label: React.Node,
  value: React.Node,
  marginRem?: number[] | number
}

class DataRowComponent extends React.PureComponent<Props & ThemeProps> {
  render() {
    const { label, value, theme } = this.props
    const styles = getStyles(theme)

    return (
      <View style={[styles.row, marginRem(this.props, theme)]}>
        <View style={styles.label}>{label}</View>
        <View style={styles.value}>{value}</View>
      </View>
    )
  }
}

function marginRem(props: Props, theme: Theme) {
  const marginRem = unpackEdges(props.marginRem || 0)

  return {
    marginBottom: theme.rem(marginRem.bottom),
    marginLeft: theme.rem(marginRem.left),
    marginRight: theme.rem(marginRem.right),
    marginTop: theme.rem(marginRem.top)
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  row: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  label: {
    flex: 1,
    marginRight: theme.rem(0.75)
  },
  value: {
    marginLeft: theme.rem(0.5),
    textAlign: 'right'
  }
}))

export const DataRow = withTheme(DataRowComponent)
