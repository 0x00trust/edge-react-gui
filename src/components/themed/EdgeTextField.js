// @flow

import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'
// $FlowFixMe = OutlinedTextField is not recognize by flow
import { type TextFieldProps, OutlinedTextField, TextField } from 'react-native-material-textfield'
import AntDesignIcon from 'react-native-vector-icons/AntDesign'

import { unpackEdges } from '../../util/edges.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'

type EdgeTextFieldProps = {
  marginRem?: number | number[]
}

type EdgeOutlinedTextFieldProps = {
  fieldRef?: ?React.ElementRef<typeof OutlinedTextField>,
  marginRem?: number | number[],
  isClearable: boolean,
  small?: boolean,
  onClear: () => void
}

class EdgeTextFieldComponent extends React.PureComponent<EdgeTextFieldProps & ThemeProps & TextFieldProps> {
  render() {
    const { theme, marginRem = 0.5, ...rest } = this.props
    const margin = unpackEdges(marginRem)
    return (
      <TextField
        containerStyle={{
          marginBottom: theme.rem(margin.bottom),
          marginLeft: theme.rem(margin.left),
          marginRight: theme.rem(margin.right),
          marginTop: theme.rem(margin.top)
        }}
        baseColor={theme.primaryText}
        errorColor={theme.dangerText}
        textColor={theme.primaryText}
        tintColor={theme.primaryText}
        {...rest}
      />
    )
  }
}

class EdgeTextFieldOutlinedComponent extends React.PureComponent<EdgeOutlinedTextFieldProps & ThemeProps & TextFieldProps> {
  clearText = () => {
    const { fieldRef, onClear } = this.props
    if (fieldRef && fieldRef.current) {
      fieldRef.current.clear()
      onClear()
    }
  }

  showSearchIcon = () => {
    const { fieldRef, value } = this.props
    if (fieldRef && fieldRef.current) {
      if (fieldRef.current.focused) return false
    }
    return !value
  }

  render() {
    const { isClearable, marginRem = 0.5, small, theme, ...rest } = this.props
    const spacings = spacingStyles(marginRem, theme)
    const styles = getStyles(theme)
    const searchContentInset = this.showSearchIcon() ? { left: theme.rem(2.25) } : null
    const contentInset = small ? { input: theme.rem(0.75), label: 0, ...searchContentInset } : searchContentInset

    return (
      <View style={styles.outlinedTextFieldContainer}>
        {this.showSearchIcon() ? (
          <View style={[styles.search, { marginLeft: spacings.marginLeft }]}>
            <AntDesignIcon name="search1" color={theme.iconDeactivated} size={theme.rem(1)} />
          </View>
        ) : null}
        <OutlinedTextField
          containerStyle={[spacings, styles.outlinedTextField]}
          contentInset={contentInset}
          baseColor={theme.secondaryText}
          errorColor={theme.dangerText}
          textColor={theme.primaryText}
          tintColor={theme.textLink}
          ref={this.props.fieldRef}
          {...rest}
        />
        {isClearable && (
          <TouchableOpacity onPress={this.clearText} style={[styles.outlinedTextFieldClearContainer, { marginRight: spacings.marginRight }]}>
            <AntDesignIcon name="close" color={theme.icon} size={theme.rem(1)} />
          </TouchableOpacity>
        )}
      </View>
    )
  }
}

function spacingStyles(margin: number | number[], theme: Theme) {
  const marginRem = unpackEdges(margin)

  return {
    marginBottom: theme.rem(marginRem.bottom),
    marginLeft: theme.rem(marginRem.left),
    marginRight: theme.rem(marginRem.right),
    marginTop: theme.rem(marginRem.top)
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  outlinedTextFieldContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    height: theme.rem(4.5),
    position: 'relative'
  },
  outlinedTextField: {
    flex: 1
  },
  outlinedTextInput: {
    paddingLeft: theme.rem(2)
  },
  outlinedTextFieldClearContainer: {
    position: 'absolute',
    right: 0,
    paddingHorizontal: theme.rem(0.75),
    justifyContent: 'center',
    paddingBottom: theme.rem(0.25) // This is needed because the OutlinedTextField also has innate padding/margin/height on the bottom
  },
  search: {
    position: 'absolute',
    left: 0,
    paddingHorizontal: theme.rem(0.75),
    paddingBottom: theme.rem(0.5)
  }
}))
export const EdgeTextField = withTheme(EdgeTextFieldComponent)
const EdgeTextFieldOutlinedInner = withTheme(EdgeTextFieldOutlinedComponent)
// $FlowFixMe = forwardRef is not recognize by flow?
export const EdgeTextFieldOutlined = React.forwardRef((props, ref) => <EdgeTextFieldOutlinedInner {...props} fieldRef={ref} />) // eslint-disable-line
