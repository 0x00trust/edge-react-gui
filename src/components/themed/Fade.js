// @flow

import * as React from 'react'
import { type AnimatedValue, Animated } from 'react-native'

type OwnProps = {
  timing?: 'ease' | 'linear',
  visible: boolean,
  hidden?: boolean,
  children: React.Node,
  onFadeoutFinish?: () => any
}
type Props = OwnProps

type LocalState = {
  visible: boolean,
  prevVisibleProp: boolean
}

class FadeComponent extends React.PureComponent<Props, LocalState> {
  animatedValue: AnimatedValue = new Animated.Value(0)
  constructor(props: Props) {
    super()
    this.animatedValue = new Animated.Value(0)
    this.state = {
      visible: props.visible,
      prevVisibleProp: false
    }
  }

  static getDerivedStateFromProps(props: Props, state: LocalState) {
    if (props.visible !== state.prevVisibleProp) return { prevVisibleProp: props.visible }

    return null
  }

  componentDidUpdate(prevProps: Props): * {
    if (this.state.prevVisibleProp !== prevProps.visible) {
      this.animate(this.state.prevVisibleProp ? 0.5 : 1)
    }
  }

  componentDidMount(): * {
    this.animate(this.props.visible ? 0.5 : 0)
  }

  animate = (toValue: number) => {
    if (toValue === 0.5) {
      this.setState({ visible: true })
    }
    Animated.timing(this.animatedValue, {
      toValue,
      duration: 500,
      useNativeDriver: true
    }).start(() => {
      this.setState({ visible: toValue === 0.5 })
      if (toValue === 1 && this.props.onFadeoutFinish) this.props.onFadeoutFinish()
    })
  }

  renderContent() {
    const { children, hidden } = this.props
    if (hidden) return children
    return this.state.visible ? children : null
  }

  render() {
    const inputRange = [0, 0.5, 1]
    const opacityAnimation = this.animatedValue.interpolate({
      inputRange,
      outputRange: [0, 1, 0]
    })

    return (
      <Animated.View
        style={{
          opacity: opacityAnimation
        }}
      >
        {this.renderContent()}
      </Animated.View>
    )
  }
}

export const Fade = FadeComponent
