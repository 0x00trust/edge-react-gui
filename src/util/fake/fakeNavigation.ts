import { NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popTo: () => {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  getState() {
    return {
      drawerOpen: false,
      index: 0,
      routes: []
    }
  }
}
