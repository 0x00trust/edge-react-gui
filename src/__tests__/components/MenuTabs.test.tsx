import { describe, expect, it } from '@jest/globals'
import { BottomTabBarProps } from '@react-navigation/bottom-tabs'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { MenuTabs } from '../../components/themed/MenuTabs'
import { fakeNavigation } from '../../util/fake/fakeNavigation'

describe('MenuTabs', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const props: BottomTabBarProps = { navigation: fakeNavigation, state: { index: 0, routes: [] } } as any

    const actual = renderer.render(<MenuTabs {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
