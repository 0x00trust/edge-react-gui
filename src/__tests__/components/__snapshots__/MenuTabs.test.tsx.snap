// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuTabs should render with loading props 1`] = `
<View>
  <DividerLine
    colors={
      Array [
        "#121d25",
        "#121d25",
      ]
    }
  />
  <LinearGradient
    colors={
      Array [
        "#121d25",
        "#121d25",
      ]
    }
    end={
      Object {
        "x": 1,
        "y": 1,
      }
    }
    start={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  />
</View>
`;
